### PR TITLE
Update .NET Core SDK

### PR DESCRIPTION
**Pull request by bricelam**
~~Blocked on dotnet/sdk#2867~~

### CODE REVIEW

**Comment by bricelam** [comment]
Desperately waiting for dotnet/arcade#811

**Comment by bricelam** [comment]
```
error NU1605: Detected package downgrade: Microsoft.NETCore.Platforms from 3.0.0-preview.19112.1 to 3.0.0-preview.19107.12. Reference the package directly from the project to select a different version.
error NU1605:  Microsoft.EntityFrameworkCore.InMemory.Tests -> Microsoft.NETCore.App 3.0.0-preview-27412-2 -> Microsoft.NETCore.Platforms (>= 3.0.0-preview.19112.1)
error NU1605:  Microsoft.EntityFrameworkCore.InMemory.Tests -> Microsoft.NETCore.Platforms (>= 3.0.0-preview.19107.12)
```

☹️ I thought I fixed this with [these MSBuild properties](https://github.com/aspnet/EntityFrameworkCore/blob/91fd469887e03b5f9c6f1775bcf1dabb5d667e7e/Directory.Build.targets#L5-L6). Any insights, @natemcmaster ? Diamond dependency involving dotnet/corefx and dotnet/core-setup?

File: global.json [hunk 1]
```diff
@@ -1,9 +1,9 @@
 {
   "tools": {
-    "dotnet": "3.0.100-preview-010010"
+    "dotnet": "3.0.100-preview4-010309"
```
Comment (natemcmaster) [inline global.json]: I would recommend not proceeding beyond preview 3 yet. Our master branch is still at preview 3. Preview 3 SDKs can be found in the 'release/3.0.1xx' column on this page: https://github.com/dotnet/core-sdk/tree/release/3.0.1xx#installers-and-binaries

**Comment by natemcmaster** [comment]
@bricelam I don't think it's a diamond dependency issue. The error says you have a forced downgrade from `Microsoft.NETCore.App 3.0.0-preview-27412-2`. This version doesn't match what is in eng/Versions.props: https://github.com/bricelam/EFCore/blob/7cb0a3f5c4711b88708845b73847047d18e9822b/eng/Versions.props#L50
```
 <MicrosoftNETCoreAppPackageVersion>3.0.0-preview-27407-4</MicrosoftNETCoreAppPackageVersion>
```
So, my guess is the real problem is that `MicrosoftNETCoreAppPackageVersion` isn't actually being used to control the implicit package version.

**Comment by natemcmaster** [comment]
@dsplaisted made changes recently to how `RuntimeFrameworkVersion`  works in https://github.com/dotnet/sdk/pull/2887. I wonder if this broke https://github.com/aspnet/EntityFrameworkCore/blob/91fd469887e03b5f9c6f1775bcf1dabb5d667e7e/Directory.Build.targets#L6

**Comment by dsplaisted** [comment]
@natemcmaster It looks like the changes you reference might break this, but I don't think they have flowed into a build of the .NET Core SDK yet, so I don't think they would be breaking anything yet.

**Comment by bricelam** [comment]
Ugg. They're back.
```
error NU1605: Detected package downgrade: Microsoft.NETCore.Platforms from 3.0.0-preview.19109.6 to 3.0.0-preview.19107.12. Reference the package directly from the project to select a different version.
error NU1605:  StreamingSample -> Microsoft.NETCore.App 3.0.0-preview3-27413-7 -> Microsoft.NETCore.Platforms (>= 3.0.0-preview.19109.6)
error NU1605:  StreamingSample -> Microsoft.NETCore.Platforms (>= 3.0.0-preview.19107.12)
```

**Comment by smitpatel** [comment]
`package downgrade` loves @bricelam :trollface:

**Comment by dsplaisted** [comment]
It looks like you're directly referencing the Microsoft.NETCore.Platforms package (though it might be implicit).  Is that the case?  What do you need to reference that package for?

**Comment by natemcmaster** [comment]
AspNet repos set BundledNETCorePlatformsPackageVersion to workaround https://github.com/dotnet/cli/issues/10528

https://github.com/aspnet/EntityFrameworkCore/blob/640bc8d568b8475604d121ba78aa29c594cf3611/Directory.Build.targets#L5

**Comment by dsplaisted** [comment]
@natemcmaster If you do that then you'll need to keep updating that version to make sure it's at least equal to the version referenced by Microsoft.NETCore.App.  When we switch to the .Ref package, this should no longer be necessary.

I imagine that this is needed because you're overriding the version of .NET Core that the SDK would give you by default.

**Comment by natemcmaster** [comment]
@dsplaisted yes, that's correct. 

@bricelam since I last reviewed this PR, it looks like the versions of NETCore.App and the SDK have changed. I think originally the problem was that RuntimeFrameworkVersion was not honored by the SDK, but now it appears this is the diamond dependency problem. I believe you need to consume a newer version of CoreFX packages to match the updated version of Microsoft.NETCore.App. It's possible this didn't surface until now because the version of the SDK you were using didn't use the BundledNETCorePlatformsPackageVersion property.

In EF Core:

https://github.com/aspnet/EntityFrameworkCore/blob/640bc8d568b8475604d121ba78aa29c594cf3611/eng/Version.Details.xml#L44-L47

https://github.com/aspnet/EntityFrameworkCore/blob/640bc8d568b8475604d121ba78aa29c594cf3611/eng/Version.Details.xml#L70-L73

In core-setup:

https://github.com/dotnet/core-setup/blob/1c23cf9539146805f8c8072c6f9f4d194f13f6bf/eng/Version.Details.xml#L8-L11
```xml
    <Dependency Name="Microsoft.NETCore.Platforms" Version="3.0.0-preview.19109.6">
      <Uri>https://github.com/dotnet/corefx</Uri>
      <Sha>0abec4390b30fdda97dc496594f9b1f9c9b20e17</Sha>
    </Dependency>
```

**Comment by bricelam** [comment]
@ajcvickers Latest errors:
```
error NU1605: Detected package downgrade: Microsoft.Win32.Primitives from 4.3.0 to 4.0.1.
error NU1605:  NETStandard.Library 1.6.0 -> System.Net.Primitives 4.0.11 -> runtime.win.System.Net.Primitives 4.3.0 -> Microsoft.Win32.Primitives (>= 4.3.0)
error NU1605:  NETStandard.Library 1.6.0 -> Microsoft.Win32.Primitives (>= 4.0.1)

error NU1605: Detected package downgrade: System.Net.Primitives from 4.3.0 to 4.0.11.
error NU1605:  NETStandard.Library 1.6.0 -> System.Net.Sockets 4.1.0 -> runtime.win.System.Net.Sockets 4.3.0 -> System.Net.Primitives (>= 4.3.0)
error NU1605:  NETStandard.Library 1.6.0 -> System.Net.Primitives (>= 4.0.11)
```

**Comment by dougbu** [comment]
> Latest errors&hellip;

I suspect there's a race between when the CI jobs restore packages and when NuGet / MyGet / whatever finish indexing. Have you tried `Retry` on the pipelines page?

**Comment by bricelam** [comment]
@dougbu This happens locally too. No idea what's actually conflicting with what. The error makes it sound like NETStandard.Library version 1.6.0 is internally inconsistent, but that's unlikely.

**Comment by bricelam** [comment]
Something may be upgrading the runtime.win.* packages to an incompatible version. Maybe Microsoft.NETCore.App...

**Comment by dsplaisted** [comment]
Can you switch to NETStandard.Library 1.6.1?  I think that should resolve the downgrades.

I think what is happening here is that the Microsoft.NETCore.App 1.1.x package adds RID-specific dependencies which end up looking like they come from NETStandard.Library and thus it looks inconsistent.  This was a problem in .NET Core 1.x, and the best solution we came up with was to always use version 1.6.1 of the NETStandard.Library package and not to use 1.6.0.

**Comment by bricelam** [comment]
That worked. New (unrelated) errors:

```
error NU1603: dotnet-ef depends on runtime.win-x64.Microsoft.WindowsDesktop.App (>= 3.0.0-preview4-27504-2) but runtime.win-x64.Microsoft.WindowsDesktop.App 3.0.0-preview4-27504-2 was not found. An approximate best match of runtime.win-x64.Microsoft.WindowsDesktop.App 3.0.0-preview4-27504-20 was resolved.
error NU1603: dotnet-ef depends on runtime.win-x86.Microsoft.WindowsDesktop.App (>= 3.0.0-preview4-27504-2) but runtime.win-x86.Microsoft.WindowsDesktop.App 3.0.0-preview4-27504-2 was not found. An approximate best match of runtime.win-x86.Microsoft.WindowsDesktop.App 3.0.0-preview4-27504-20 was resolved.
error NU1102: Unable to find package Microsoft.AspNetCore.App.Runtime.win-x64 with version (>= 3.0.0-preview4-27504-2)
error NU1102: Unable to find package Microsoft.AspNetCore.App.Runtime.win-x86 with version (>= 3.0.0-preview4-27504-2)
```

@wli3 This is happening in the dotnet-ef project. Why is it even trying to restore these?

**Comment by dsplaisted** [comment]
@bricelam When targeting .NET Core 3.0, currently it will restore all of the targeting packs (which includes WindowsDesktop) if they aren't part of the SDK.  This is because (eventually) there could be a transitive FrameworkReference to that shared framework which is discovered after restore.

Probably the problem here is that you are overriding `RestoreAdditionalProjectSources` instead of appending to it.  If you don't override it, then it should contain the feeds you need to restore that package.  See https://github.com/dotnet/core-sdk/blob/master/src/redist/targets/GenerateBundledVersions.targets#L169

**Comment by bricelam** [comment]
Hmmm, we're just using Arcade's `RestoreSources` property. I can't see any interaction between Arcade and `RestoreAdditionalProjectSources`...

**Comment by bricelam** [comment]
Also, https://dotnetfeed.blob.core.windows.net/dotnet-core/index.json is being used as a source.

**Comment by dsplaisted** [comment]
Sorry, there is a whole list of default sources, and I just linked the first line.  The link should have been https://github.com/dotnet/core-sdk/blob/master/src/redist/targets/GenerateBundledVersions.targets#L169-L174

For this package, you probably need https://dotnetfeed.blob.core.windows.net/dotnet-windowsdesktop/index.json

**Comment by bricelam** [comment]
Might actually be a publishing issue. I don't see version 3.0.0-preview4-27504-2 on the feed.

**Comment by dsplaisted** [comment]
You are right, that is not the correct version of the package that it should be using.

Are you setting the `RuntimeFrameworkVersion`?  That is a hammer that now applies to all shared frameworks, which doesn't really work when we're in previews and each of the shared frameworks have a different version.

**Comment by bricelam** [comment]
Yes. We [set it](https://github.com/aspnet/EntityFrameworkCore/blob/master/Directory.Build.targets#L6) to dogfood the latest Microsoft.NETCore.App. Is there an alternative? Should even be trying to dogfood it outside of the SDK?

**Comment by dsplaisted** [comment]
I don't know whether you should be dogfooding it outside of the SDK.  It certainly has the potential to make things more complicated.

However, if you do want to just use the specified version of Microsoft.NETCore.App, then you should be able to do the following:

```xml
<ItemGroup>
    <FrameworkReference Update="Microsoft.NETCore.App" RuntimeFrameworkVersion="$(MicrosoftNETCoreAppPackageVersion)" />
</ItemGroup>
```

IE, it is possible to specify the `RuntimeFrameworkVersion` per FrameworkReference, and that code adds that metadata to the implicit FrameworkReference for Microsoft.NETCore.App.

**Comment by bricelam** [comment]
Things are currently way to complex to manage. Going to try and reduce our direct dependencies to just the ones that are actually needed.

**Comment by bricelam** [comment]
New error
```
error MSB4018: The "GenerateShims" task failed unexpectedly.
error MSB4018: System.InvalidOperationException: Sequence contains no matching element
error MSB4018:    at System.Linq.ThrowHelper.ThrowNoMatchException()
error MSB4018:    at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source, Func`2 predicate)
error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateShims.GetApphostAsset(ITaskItem[] apphostsForShimRuntimeIdentifiers, String runtimeIdentifier) in /_/src/Tasks/Microsoft.NET.Build.Tasks/GenerateShims.cs:line 121
error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateShims.ExecuteCore() in /_/src/Tasks/Microsoft.NET.Build.Tasks/GenerateShims.cs:line 75
error MSB4018:    at Microsoft.NET.Build.Tasks.TaskBase.Execute() in /_/src/Tasks/Common/TaskBase.cs:line 35
error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask)
```

wack-a-mole

**Comment by dsplaisted** [comment]
The GenerateShims failure is actually a bug I fixed yesterday with https://github.com/dotnet/sdk/pull/3002

Unfortunately it looks like it hasn't flowed into a build of the .NET Core SDK.  I'll follow up on that.

File: eng/Versions.props [hunk 2]
```diff
@@ -27,26 +27,30 @@
...
   </PropertyGroup>
   <PropertyGroup Label="Dependencies from dotnet/core-setup">
-    <MicrosoftExtensionsDependencyModelPackageVersion>3.0.0-preview4-27501-1</MicrosoftExtensionsDependencyModelPackageVersion>
-    <MicrosoftNETCoreAppPackageVersion>3.0.0-preview4-27501-1</MicrosoftNETCoreAppPackageVersion>
+    <MicrosoftDotNetPlatformAbstractionsPackageVersion>3.0.0-preview4-27504-4</MicrosoftDotNetPlatformAbstractionsPackageVersion>
+    <MicrosoftExtensionsDependencyModelPackageVersion>3.0.0-preview4-27504-4</MicrosoftExtensionsDependencyModelPackageVersion>
+    <MicrosoftNETCoreAppPackageVersion>3.0.0-preview4-27504-4</MicrosoftNETCoreAppPackageVersion>
+  </PropertyGroup>
+  <PropertyGroup Label="Dependencies from dotnet/standard">
+    <NETStandardLibraryPackageVersion>2.1.0-prerelease.19126.1</NETStandardLibraryPackageVersion>
```
Comment (dougbu) [inline eng/Versions.props]: Did you add a subscription from dotnet/standard to aspnet/EntityFrameworkCore in the `'.NET Core 3 Dev'` channel? Without that, this line will only ever get updated manually.

File: eng/common/darc-init.ps1 [hunk 3]
```diff
@@ -19,7 +19,7 @@ function InstallDarcCli ($darcVersion) {
   # Until we can anonymously query the BAR API for the latest arcade-services
   # build applied to the PROD channel, this is hardcoded.
   if (-not $darcVersion) {
-    $darcVersion = '1.1.0-beta.19120.2'
+    $darcVersion = '1.1.0-beta.19151.3'
```
Comment (dougbu) [inline eng/common/darc-init.ps1]: Do _not_ update files in `eng/common` manually. These changes will be overwritten the next time we update the Arcade SDK.

FYI this is only a default. Can pass whatever version you like on the command line.

Comment (bricelam) [reply eng/common/darc-init.ps1]: I didn't. darc did.

File: eng/Versions.props (hunk 2 above)
Comment (bricelam) [reply eng/Versions.props]: It's on my TODO list.

**Comment by dougbu** [comment]
Of course, the macOS failure may indicate an issue ☹️

File: eng/Versions.props (hunk 2 above)
Comment (bricelam) [reply eng/Versions.props]: Removed. Not needed.

**Comment by bricelam** [comment]
Thanks for your help yesterday, @dsplaisted. It looks like we're finally unblocked.

**Comment by bricelam** [comment]
macOS failure looks like some race condition in MSBuild or Arcade. Unrelated to my changes.

Removing some unnecessary dependencies. We'll see if it fails again.

File: global.json [hunk 4]
```diff
@@ -1,9 +1,9 @@
 {
   "tools": {
-    "dotnet": "3.0.100-preview-010010"
+    "dotnet": "3.0.100-preview4-010612"
```
Comment (bricelam) [inline global.json]: @roji Hopefully in the future, updating the SDK will only require changing this file. 😉

**Comment by roji** [comment]
@bricelam you're my hero...

**Pull request merged**